File: Frameworks/react-web/tests/unit/models/ods-visible-when.test.ts
import { describe, it, expect } from 'vitest'
import {
  parseComponentVisibleWhen,
  parseVisibleWhen,
  isFieldBased,
  isDataBased,
  type OdsComponentVisibleWhen,
  type OdsVisibleWhen,
} from '../../../src/models/ods-visible-when.ts'

// ===========================================================================
// OdsComponentVisibleWhen model tests
// ===========================================================================

describe('parseComponentVisibleWhen', () => {
  // -------------------------------------------------------------------------
  // Null / undefined input
  // -------------------------------------------------------------------------

  it('returns undefined for null input', () => {
    expect(parseComponentVisibleWhen(null)).toBeUndefined()
  })

  it('returns undefined for undefined input', () => {
    expect(parseComponentVisibleWhen(undefined)).toBeUndefined()
  })

  it('returns undefined for non-object input', () => {
    expect(parseComponentVisibleWhen('string')).toBeUndefined()
  })

  it('returns undefined for number input', () => {
    expect(parseComponentVisibleWhen(42)).toBeUndefined()
  })

  // -------------------------------------------------------------------------
  // Field-based condition
  // -------------------------------------------------------------------------

  it('parses field-based condition', () => {
    const result = parseComponentVisibleWhen({
      field: 'status',
      form: 'mainForm',
      equals: 'active',
    })
    expect(result).toBeDefined()
    expect(result!.field).toBe('status')
    expect(result!.form).toBe('mainForm')
    expect(result!.equals).toBe('active')
  })

  it('parses field-based condition with notEquals', () => {
    const result = parseComponentVisibleWhen({
      field: 'role',
      form: 'userForm',
      notEquals: 'guest',
    })
    expect(result!.field).toBe('role')
    expect(result!.form).toBe('userForm')
    expect(result!.notEquals).toBe('guest')
    expect(result!.equals).toBeUndefined()
  })

  // -------------------------------------------------------------------------
  // Data-based condition
  // -------------------------------------------------------------------------

  it('parses data-based condition with source and countEquals', () => {
    const result = parseComponentVisibleWhen({
      source: 'tasks',
      countEquals: 0,
    })
    expect(result!.source).toBe('tasks')
    expect(result!.countEquals).toBe(0)
  })

  it('parses data-based condition with countMin and countMax', () => {
    const result = parseComponentVisibleWhen({
      source: 'items',
      countMin: 1,
      countMax: 10,
    })
    expect(result!.source).toBe('items')
    expect(result!.countMin).toBe(1)
    expect(result!.countMax).toBe(10)
  })

  it('parses data-based condition with only source', () => {
    const result = parseComponentVisibleWhen({ source: 'records' })
    expect(result!.source).toBe('records')
    expect(result!.countEquals).toBeUndefined()
    expect(result!.countMin).toBeUndefined()
    expect(result!.countMax).toBeUndefined()
  })

  // -------------------------------------------------------------------------
  // equals coercion via String()
  // -------------------------------------------------------------------------

  it('coerces numeric equals to string', () => {
    const result = parseComponentVisibleWhen({
      field: 'count',
      form: 'f',
      equals: 42,
    })
    expect(result!.equals).toBe('42')
  })

  it('coerces boolean equals to string', () => {
    const result = parseComponentVisibleWhen({
      field: 'active',
      form: 'f',
      equals: true,
    })
    expect(result!.equals).toBe('true')
  })

  it('coerces numeric notEquals to string', () => {
    const result = parseComponentVisibleWhen({
      field: 'count',
      form: 'f',
      notEquals: 0,
    })
    expect(result!.notEquals).toBe('0')
  })

  it('leaves equals undefined when not provided', () => {
    const result = parseComponentVisibleWhen({
      field: 'x',
      form: 'f',
    })
    expect(result!.equals).toBeUndefined()
    expect(result!.notEquals).toBeUndefined()
  })

  // -------------------------------------------------------------------------
  // Empty object
  // -------------------------------------------------------------------------

  it('parses empty object with all fields undefined', () => {
    const result = parseComponentVisibleWhen({})
    expect(result).toBeDefined()
    expect(result!.field).toBeUndefined()
    expect(result!.form).toBeUndefined()
    expect(result!.source).toBeUndefined()
  })
})

// ===========================================================================
// isFieldBased / isDataBased type guards
// ===========================================================================

describe('isFieldBased', () => {
  it('returns true when field and form are set', () => {
    const v: OdsComponentVisibleWhen = { field: 'status', form: 'mainForm', equals: 'yes' }
    expect(isFieldBased(v)).toBe(true)
  })

  it('returns false when field is missing', () => {
    const v: OdsComponentVisibleWhen = { form: 'mainForm' }
    expect(isFieldBased(v)).toBe(false)
  })

  it('returns false when form is missing', () => {
    const v: OdsComponentVisibleWhen = { field: 'status' }
    expect(isFieldBased(v)).toBe(false)
  })

  it('returns false when both field and form are missing', () => {
    const v: OdsComponentVisibleWhen = { source: 'tasks' }
    expect(isFieldBased(v)).toBe(false)
  })
})

describe('isDataBased', () => {
  it('returns true when source is set', () => {
    const v: OdsComponentVisibleWhen = { source: 'tasks', countEquals: 0 }
    expect(isDataBased(v)).toBe(true)
  })

  it('returns false when source is not set', () => {
    const v: OdsComponentVisibleWhen = { field: 'status', form: 'f' }
    expect(isDataBased(v)).toBe(false)
  })

  it('returns false for empty object', () => {
    const v: OdsComponentVisibleWhen = {}
    expect(isDataBased(v)).toBe(false)
  })
})

// ===========================================================================
// OdsVisibleWhen (field-level) model tests
// ===========================================================================

describe('parseVisibleWhen', () => {
  it('parses valid input with field and equals', () => {
    const result = parseVisibleWhen({ field: 'type', equals: 'advanced' })
    expect(result).toBeDefined()
    expect(result!.field).toBe('type')
    expect(result!.equals).toBe('advanced')
  })

  it('returns undefined for null input', () => {
    expect(parseVisibleWhen(null)).toBeUndefined()
  })

  it('returns undefined for undefined input', () => {
    expect(parseVisibleWhen(undefined)).toBeUndefined()
  })

  it('returns undefined for non-object input', () => {
    expect(parseVisibleWhen('not-object')).toBeUndefined()
  })

  it('returns undefined for number input', () => {
    expect(parseVisibleWhen(123)).toBeUndefined()
  })

  it('parses object with missing equals', () => {
    const result = parseVisibleWhen({ field: 'status' })
    expect(result).toBeDefined()
    expect(result!.field).toBe('status')
    expect(result!.equals).toBeUndefined()
  })

  it('parses object with missing field', () => {
    const result = parseVisibleWhen({ equals: 'yes' })
    expect(result).toBeDefined()
    expect(result!.field).toBeUndefined()
    expect(result!.equals).toBe('yes')
  })
})

File: Frameworks/flutter-local/test/models/ods_visible_when_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:ods_flutter_local/models/ods_visible_when.dart';

void main() {
  // =========================================================================
  // fromJson
  // =========================================================================

  group('OdsComponentVisibleWhen.fromJson', () {
    test('parses field-based condition with equals', () {
      final vw = OdsComponentVisibleWhen.fromJson({
        'field': 'type',
        'form': 'myForm',
        'equals': 'Other',
      });
      expect(vw.field, 'type');
      expect(vw.form, 'myForm');
      expect(vw.equals, 'Other');
      expect(vw.notEquals, isNull);
      expect(vw.source, isNull);
    });

    test('parses field-based condition with notEquals', () {
      final vw = OdsComponentVisibleWhen.fromJson({
        'field': 'status',
        'form': 'taskForm',
        'notEquals': 'completed',
      });
      expect(vw.field, 'status');
      expect(vw.form, 'taskForm');
      expect(vw.notEquals, 'completed');
      expect(vw.equals, isNull);
    });

    test('parses data-based condition', () {
      final vw = OdsComponentVisibleWhen.fromJson({
        'source': 'quiz_answers',
        'countEquals': 10,
        'countMin': 5,
        'countMax': 20,
      });
      expect(vw.source, 'quiz_answers');
      expect(vw.countEquals, 10);
      expect(vw.countMin, 5);
      expect(vw.countMax, 20);
      expect(vw.field, isNull);
      expect(vw.form, isNull);
    });

    test('all fields null by default', () {
      final vw = OdsComponentVisibleWhen.fromJson({});
      expect(vw.field, isNull);
      expect(vw.form, isNull);
      expect(vw.equals, isNull);
      expect(vw.notEquals, isNull);
      expect(vw.source, isNull);
      expect(vw.countEquals, isNull);
      expect(vw.countMin, isNull);
      expect(vw.countMax, isNull);
    });
  });

  // =========================================================================
  // isFieldBased
  // =========================================================================

  group('isFieldBased', () {
    test('returns true when field AND form are set', () {
      final vw = OdsComponentVisibleWhen.fromJson({
        'field': 'category',
        'form': 'mainForm',
        'equals': 'premium',
      });
      expect(vw.isFieldBased, isTrue);
    });

    test('returns false when only field is set', () {
      final vw = OdsComponentVisibleWhen.fromJson({
        'field': 'category',
      });
      expect(vw.isFieldBased, isFalse);
    });

    test('returns false when only form is set', () {
      final vw = OdsComponentVisibleWhen.fromJson({
        'form': 'mainForm',
      });
      expect(vw.isFieldBased, isFalse);
    });
  });

  // =========================================================================
  // isDataBased
  // =========================================================================

  group('isDataBased', () {
    test('returns true when source is set', () {
      final vw = OdsComponentVisibleWhen.fromJson({
        'source': 'answers',
        'countMin': 1,
      });
      expect(vw.isDataBased, isTrue);
    });

    test('returns false when source is null', () {
      final vw = OdsComponentVisibleWhen.fromJson({
        'field': 'type',
        'form': 'myForm',
      });
      expect(vw.isDataBased, isFalse);
    });
  });
}
